### ISSUE TITLE
Request to Add Additional Data Augmentation Techniques

### ISSUE THREAD

**Issue by alvarovegahd** [shyamblast/Koogu#9]
Hello!

First of all, thank you for developing such a useful tool.

It would be fantastic to include additional data augmentation techniques in the package, beyond the already implemented [Gaussian Noise and Volume Ramping](https://shyamblast.github.io/Koogu/en/stable/advanced/data_augmentation.html). Specifically, I'd love to see Echo, Background Infuse, Variable Noise Floor, and Alter Distance as described in [this publication](https://royalsocietypublishing.org/doi/pdf/10.1098/rstb.2023.0444). Are there any plans to add these, or could you offer advice on how to contribute these new features?

All the best,

**Comment by NickleDave** [comment]
I'm not @shyamblast but I'm pretty sure these are already integrated as described in the "data accessibilty" section of the article you linked.

> Data considered in this study were from previously published works [33,50]. The techniques developed in this study have been integrated into recent releases of the open-source Koogu toolbox: https://github.com/shyamblast/Koogu. The final trained model is also bundled with releases (starting from v.1.6.5) of Raven Pro software. For replication of the results presented in the manuscript, all training audio (focal, in situ and playback) and annotations (in Raven Pro selection table format) can be obtained from the Zenodo Repository: https://doi.org/10. 5281/zenodo.10594015 [74] and all testing audio files and annotations (also Raven Pro format) are available from the Dryad Digital Repository:
https://doi.org/10.5061/dryad.zw3r2288b [75].

Looks like they are here in the docs: https://shyamblast.github.io/Koogu/en/stable/api/data/augmentation.html#time-domain-augmentations and https://shyamblast.github.io/Koogu/en/stable/api/data/augmentation.html#spectro-temporal-augmentations. Not sure if there's any example scripts for how to use them tho'

**Comment by NickleDave** [comment]
FWIW you might also find this related work interesting
- https://www.sciencedirect.com/science/article/pii/S1574954124001158?via%3Dihub
- https://github.com/houtan-ghaffari/bird-frontends

**Comment by shyamblast** [comment]
Thanks for the quick help @NickleDave 

@alvarovegahd , as Dave mentioned, implementations of most of the referenced [pre-transform](https://shyamblast.github.io/Koogu/en/stable/api/data/augmentation.html#time-domain-augmentations) (in time-domain) and [post-transform](https://shyamblast.github.io/Koogu/en/stable/api/data/augmentation.html#spectro-temporal-augmentations) (in spectrotemporal domain) augmentations are already available in Koogu ("transform" simply means converting audio waveforms to spectrograms). Examples for using these API interfaces are provided [here](https://shyamblast.github.io/Koogu/en/stable/advanced/data_augmentation.html). Note that, as of this writing, you'd need to extend one of the base classs ([BaseFeeder](https://shyamblast.github.io/Koogu/en/stable/api/data/feeder.html#koogu.data.feeder.BaseFeeder) or [SpectralBaseFeeder](https://shyamblast.github.io/Koogu/en/stable/api/data/feeder.html#koogu.data.feeder.SpectralDataFeeder)) in order to include augmentations. As for the referenced publication, I had inherited from the BaseFeeder, since I needed to include Background Infuse and Variable Noise Floor augmentations (given that different bioacoustics ML projects manage/handle data in different ways, implementing API interfaces for these two augmentations isn't straightforward).

The below function can be used for infusing background noise (per specifications mentioned in the published article). `clip` and `bg` parameters refer to TF arrays corresponding to foreground signal waveform and background noise waveform, respectively.

    def scaled_background(clip, bg):
        
        # Alter the background (a bit) too, and then re-normalize
        bg = Temporal.RampVolume((-9, 9))(bg)
        bg = bg / tf.reduce_max(tf.abs(bg), axis=-1, keepdims=True)
        
        # Attenuate 'background' to a randomized -dB peak level
        randomizer_args = (-18, -12, tf.float32)     # dB
        return bg * (
            tf.pow(10.0, (tf.random.uniform([], *randomizer_args) / 20.0)) *
            tf.reduce_max(tf.abs(clip))  # scale relative to peak
        )
The array returned by `scaled_background()` must be added to the original clip array. Do these within the `pre_transform()` function in your Feeder implementation.

As for variable noise floor, you can insert the below code (with appropriate alterations to the noise levels) in the `transform()` function of your Feeder implementation.

      output = output / tf.reduce_max(tf.abs(output),
                                      axis=-1, keepdims=True)

      # Convert to spectrogram, with varying noise floor.
      # set the noise floor to be a random value between -105 dB to -85 dB
      noise_floor = tf.math.pow(10.0,
                                tf.random.uniform([], -105.0, -85.0) / 10.0)
      output = Audio2Spectral(self._fs, self._spec_settings,
                              eps=noise_floor)(output)
